File: week_0/25979.py
import sys
def cal_idx(time):
    h, m, s = map(int, time.split(':'))
    return (((h*60)+m)*60)+s

def add_arr(s_idx, e_idx):
    t_arr[s_idx] += 1
    t_arr[e_idx] += 1

def cal_prefix_sum():
    v = t_arr[0]
    prefix_sum[0] = v
    for i in range(1, 86398):
        v += t_arr[i]
        prefix_sum[i] = prefix_sum[i-1] + v

def find_max(len):
    max = prefix_sum[len-1]
    cal_prefix_sum()
    for i in range(0, max_idx-len+1):
        sum = prefix_sum[len+i-1] - prefix_sum[i-1]
        if sum > max:
            max = sum
            print(f'max = {max}, {i} ~ {len+i}')
    return max

t_arr = [0] * 86400
prefix_sum = [0] * 86400
min_idx = 86399
max_idx = 0
num = int(sys.stdin.readline())

for i in range(num):
    cmd = sys.stdin.readline()
    if int(cmd[0]) == 1:
        a, b, c = cmd.split()
        start_idx = cal_idx(b)
        end_idx = cal_idx(c)
        if start_idx < min_idx:
            min_idx = start_idx
        if end_idx > max_idx:
            max_idx = end_idx
        add_arr(start_idx, end_idx)
    else:
        a, b = cmd.split()
        c = cal_idx(b)
        print(find_max(c))

File: 25979.c
#include <stdio.h>
#include <string.h>

double in_arr[86400];
double prefix_sum[86400];

int cal_idx(int h, int m, int s)
{
    return (3600*h + 60*m + s);
}

void add_arr(int s_idx, int e_idx)
{
    in_arr[s_idx]+=1;
    if (e_idx < 86400)
        in_arr[e_idx]-=1;
}

void cal_prefix_sum()
{
    double v = in_arr[0];
    prefix_sum[0] = in_arr[0];
    for(int i = 1; i < 86400; i++){
        v += in_arr[i];
        prefix_sum[i] = prefix_sum[i-1] + v;
    }
}

double find_max(int len)
{
    double max = prefix_sum[len-1];
    double sum;
    cal_prefix_sum();
    for (int i = 1; i < 86400-len+1; i++)
    {
        sum = prefix_sum[len+i-1] - prefix_sum[i-1];
        if (sum > max){
            max = sum;
        }
    }
    return max;
}

int main(void)
{
    int num, start_idx, end_idx, cmd;
    scanf("%d", &num);
    memset(in_arr, 0, sizeof(double)*86400);
    for (int i = 1; i < num; i++)
    {
        scanf("%d", &cmd);
        if (cmd == 1)
        {
            int h1, h2, m1, m2, s1, s2;
            scanf("%d:%d:%d %d:%d:%d", &h1, &m1, &s1, &h2, &m2, &s2);
            start_idx = cal_idx(h1, m1, s1);
            end_idx = cal_idx(h2, m2, s2);
            add_arr(start_idx, end_idx);
        }
    }
    scanf("%d", &cmd);
    if (cmd == 2)
    {
        int h3, m3, s3, t;
        scanf("%d:%d:%d", &h3, &m3, &s3);
        t = cal_idx(h3, m3, s3);
        printf("%.0lf\n", find_max(t));
    }
    return 0;
}
